### PR TITLE
feat: support max timestamp stats for data skipping

### PR DESCRIPTION
**Pull request by DrakeLin**
## 🥞 Stacked PR
Use this [link](https://github.com/delta-io/delta-kernel-rs/pull/2249/files) to review incremental changes.
- [**stack/support-max-timestamp-pruning**](https://github.com/delta-io/delta-kernel-rs/pull/2249) [[Files changed](https://github.com/delta-io/delta-kernel-rs/pull/2249/files)]

---------
## What changes are proposed in this pull request?

Enables data skipping on timestamp columns' max stats, which were previously disabled entirely due to Delta JSON stats truncating timestamps to millisecond precision. The fix subtracts 999us from the comparison literal before comparing against the max stat, accounting for the worst-case truncation error.

## How was this change tested?
New unit and integration tests, changed existing unit tests

Tested against a Spark-written Delta table with timestamps. Read with scan_metadata with equality predicate to verify pruning

### CODE REVIEW

File: kernel/src/scan/data_skipping.rs [hunk 1]
```diff
@@ -553,14 +586,11 @@ impl DataSkippingPredicateEvaluator for NullGuardedDataSkippingPredicateCreator<
...
-    fn get_max_stat(&self, col: &ColumnName, data_type: &DataType) -> Option<Expr> {
+    fn get_max_stat(&self, col: &ColumnName, _data_type: &DataType) -> Option<Expr> {
         if self.is_partition_column(col) {
             return None;
         }
-        match data_type {
-            &DataType::TIMESTAMP | &DataType::TIMESTAMP_NTZ => None,
-            _ => Some(joined_column_expr!("maxValues", col)),
-        }
+        Some(joined_column_expr!("maxValues", col))
```
Comment (dengsh12) [inline kernel/src/scan/data_skipping.rs:593]: Do we want to just delete the para `_data_type`?

**Review by dengsh12 — APPROVED** [review]
LGTM ~assuming it's a mid-term fix. However, as it's a reader-side fix not writer-side, wonder if we want a long term fix on writer side? I.e. after this fix, if other engines are reading kernel-written tables, they will still incorrectly skip data bc timestamp truncating (:~ Discussed offline, learned kernel writes timestamp stats correctly, this is for compatible with spark

File: kernel/src/scan/data_skipping.rs [hunk 2]
```diff
@@ -370,6 +371,23 @@ fn collect_junction_preds(
     Pred::junction(op, preds)
 }
 
+/// Adjusts a comparison value before comparing against a max stat, to account for existing
+/// Delta writers truncating timestamp JSON stats to millisecond precision. Truncation floors
+/// to the millisecond, so the stored max can be up to 999us less than the actual maximum
```
Comment (sanujbasu) [inline kernel/src/scan/data_skipping.rs]: we can reference the Delta protocol directly here since 999us i.e milliseconds stats precision is 
enforced in the spec. Eg:

```
  /// Adjusts a comparison value before comparing against a max stat, to account for the Delta
  /// protocol requiring timestamp stats to be truncated to millisecond precision (see the
  /// Per-file Statistics spec footnote: "Timestamp columns are truncated down to milliseconds").
  /// Truncation floors to the millisecond, so the stored max can be up to 999us less than the
  /// actual maximum (`stored_max <= actual_max <= stored_max + 999`). We subtract 999us from
  /// the comparison value to avoid incorrectly pruning files.
  ///
  /// Non-timestamp values pass through unchanged.
```

File: kernel/src/scan/data_skipping.rs [hunk 3]
```diff
@@ -370,6 +371,23 @@ fn collect_junction_preds(
...
+/// (`actual_max <= stored_max + 999`). We subtract 999us from the comparison value to avoid
+/// incorrectly pruning files.
+///
+/// Non-timestamp values pass through unchanged.
+fn adjust_scalar_for_max_stat_truncation(val: &Scalar) -> Cow<'_, Scalar> {
+    match val {
+        Scalar::Timestamp(micros) => Cow::Owned(Scalar::Timestamp(micros.saturating_sub(999))),
+        Scalar::TimestampNtz(micros) => {
+            Cow::Owned(Scalar::TimestampNtz(micros.saturating_sub(999)))
```
Comment (sanujbasu) [inline kernel/src/scan/data_skipping.rs]: since this is the tricky part of the PR add an example?

File: kernel/src/scan/data_skipping.rs (hunk 1 above)
Comment (sanujbasu) [reply kernel/src/scan/data_skipping.rs:593]: +1 . Isn't this a dead argument now?

File: kernel/src/scan/data_skipping/tests.rs [hunk 4]
```diff
@@ -497,48 +526,115 @@ fn test_checkpoint_skipping_conjunction_partial_null_stats() {
...
+    for timestamp in [
+        &Scalar::Timestamp(1_000_000),
+        &Scalar::TimestampNtz(1_000_000),
+    ] {
+        // GT: should produce OR(maxValues.ts_col IS NULL, maxValues.ts_col > 999001)
+        let pred = Pred::gt(col.clone(), timestamp.clone());
+        let skipping_pred = as_checkpoint_skipping_predicate(&pred, &[]).unwrap();
+        let s = skipping_pred.to_string();
+        assert!(
+            s.contains("999001"),
```
Comment (sanujbasu) [inline kernel/src/scan/data_skipping/tests.rs]: why not use 

```
        assert_eq!(
            skipping_pred.unwrap().to_string(),

```
as below?

File: kernel/src/scan/data_skipping.rs [hunk 5]
```diff
@@ -370,6 +371,23 @@ fn collect_junction_preds(
     Pred::junction(op, preds)
 }
 
+/// Adjusts a comparison value before comparing against a max stat, to account for existing
+/// Delta writers truncating timestamp JSON stats to millisecond precision. Truncation floors
+/// to the millisecond, so the stored max can be up to 999us less than the actual maximum
+/// (`actual_max <= stored_max + 999`). We subtract 999us from the comparison value to avoid
```
Comment (sanujbasu) [inline kernel/src/scan/data_skipping.rs]: stored_max <= actual_max <= stored_max + 999

File: kernel/tests/read.rs [hunk 6]
```diff
@@ -2068,3 +2070,165 @@ fn partition_values_parsed_skipping() -> Result<(), Box<dyn std::error::Error>>
...
+//   file 1: ts_col [1s, 2s]           -- max at ms boundary
+//   file 2: ts_col [3s, 4.000500s]    -- JSON max truncated to 4.000s
+//   file 3: ts_col [7s, 8s]           -- max at ms boundary
+//
+// Predicate `ts_col > 4_000_400us`:
+//   file 1: max=2s << adjusted predicate (3_999_401) -> pruned (skipping works)
+//   file 2: truncated max=4s > adjusted predicate (3_999_401) -> kept (truncation safe)
+//   file 3: max=8s >> adjusted predicate -> kept
+#[tokio::test]
+async fn timestamp_max_stat_truncation_does_not_over_prune(
```
Comment (sanujbasu) [inline kernel/tests/read.rs:2084]: can we make the integration tests richer to cover all different operators?

```
  Integration test vs unit test operator coverage

  ┌──────────┬────────────────────────────────────────────────────────────┬──────────────────────────────────────────────────────┐
  │ Operator │                         Unit test                          │                   Integration test                   │
  │          │     (test_timestamp_predicates_use_adjusted_max_stats)     │                                                      │
  ├──────────┼────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────┤
  │ LT       │                            Yes                             │                Yes (bare_ts_lt_skips)                │
  ├──────────┼────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────┤
  │ GT       │                            Yes                             │  Yes (bare_ts_gt_keeps_all, in-memory test, Spark    │
  │          │                                                            │                     table test)                      │
  ├──────────┼────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────┤
  │ GE       │                            Yes                             │                          No                          │
  ├──────────┼────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────┤
  │ LE       │                            Yes                             │                          No                          │
  ├──────────┼────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────┤
  │ EQ       │                            Yes                             │                          No                          │
  ├──────────┼────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────┤
  │ NE       │                            Yes                             │                          No                          │
  └──────────┴────────────────────────────────────────────────────────────┴──────────────────────────────────────────────────────┘
```

File: kernel/tests/read.rs [hunk 7]
```diff
@@ -2068,3 +2070,165 @@ fn partition_values_parsed_skipping() -> Result<(), Box<dyn std::error::Error>>
...
+    let ts_schema = Arc::new(ArrowSchema::new(vec![ArrowField::new(
+        "ts_col",
+        delta_kernel::arrow::datatypes::DataType::Timestamp(
+            TimeUnit::Microsecond,
+            Some("UTC".into()),
+        ),
+        true,
+    )]));
+
+    // file1: max at ms boundary, will be pruned
```
Comment (sanujbasu) [inline kernel/tests/read.rs:2106]: Do we look at the boundary conditions for truncation the tests?  aka

```
  ┌────────────────────────────────────────────────────────────┬───────────┬─────────────────────────┐
  │                          Scenario                          │ Unit test │    Integration test     │
  ├────────────────────────────────────────────────────────────┼───────────┼─────────────────────────┤
  │ Comparison value at exact ms boundary (e.g., 4_000_000)    │    No     │           No            │
  ├────────────────────────────────────────────────────────────┼───────────┼─────────────────────────┤
  │ Comparison value 1us above ms boundary (e.g., 4_000_001)   │    No     │           No            │
  ├────────────────────────────────────────────────────────────┼───────────┼─────────────────────────┤
  │ Comparison value 999us above ms boundary (e.g., 4_000_999) │    No     │           No            │
  ├────────────────────────────────────────────────────────────┼───────────┼─────────────────────────┤
  │ Comparison value mid-ms (e.g., 4_000_400)                  │    No     │ Yes (in-memory + Spark) │
  ├────────────────────────────────────────────────────────────┼───────────┼─────────────────────────┤
  │ Stored max == adjusted val (exact threshold)               │    No     │           No            │
  ├────────────────────────────────────────────────────────────┼───────────┼─────────────────────────┤
  │ Stored max == adjusted val - 1 (just prunable)             │    No     │           No            │
  ├────────────────────────────────────────────────────────────┼───────────┼─────────────────────────┤
  │ Stored max == adjusted val + 1 (just not prunable)         │    No     │           No            │
  └────────────────────────────────────────────────────────────┴───────────┴─────────────────────────┘
```

File: kernel/src/scan/data_skipping.rs (hunk 1 above)
Comment (DrakeLin) [reply kernel/src/scan/data_skipping.rs:593]: can't remove without changing the trait

File: kernel/src/scan/data_skipping.rs (hunk 2 above)
Comment (DrakeLin) [reply kernel/src/scan/data_skipping.rs]: Done -- updated doc comment to reference the protocol spec and fixed the invariant direction. See 5122fbc0.
